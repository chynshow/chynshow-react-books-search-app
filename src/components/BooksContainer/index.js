import React, { useCallback, useContext, useRef } from 'react';
import { AppContext } from '../../state/AppContext';
import BookItem from '../BookItem';
import TotalResult from '../TotalResult';

const BooksContainer = () => {
  const {
    books,
    loading,
    startIndex,
    maxResults,
    totalResults,
    fetchBooks,
    searchParams,
    setStartIndex,
  } = useContext(AppContext);

  // Load new items based on position of last item in the list and "totalResults" parameter
  const observer = useRef();
  const lastBookElRef = useCallback(
    (el) => {
      if (loading) return;
      if (observer.current) observer.current.disconnect();
      observer.current = new IntersectionObserver((entries) => {
        if (
          entries[0].isIntersecting &&
          startIndex + maxResults < totalResults
        ) {
          setStartIndex();
          fetchBooks(searchParams, startIndex, maxResults);
        }
      });
      if (el) observer.current.observe(el);
    },
    [
      loading,
      searchParams,
      startIndex,
      maxResults,
      totalResults,
      setStartIndex,
      fetchBooks,
    ]
  );

  return (
    <ul className='l-book-container'>
      <TotalResult />
      {
        // eslint-disable-next-line array-callback-return
        books.map(({ volumeInfo, id }, idx) => {
          if (volumeInfo && volumeInfo.imageLinks) {
            if (books.length === idx + 1) {
              return (
                <BookItem
                  title={volumeInfo.title}
                  imageLinks={volumeInfo.imageLinks}
                  description={volumeInfo.description}
                  key={id}
                  reference={lastBookElRef}
                />
              );
            } else {
              return (
                <BookItem
                  key={id}
                  title={volumeInfo.title}
                  imageLinks={volumeInfo.imageLinks}
                  description={volumeInfo.description}
                />
              );
            }
          }
        })
      }
      {loading && <div className='c-loader'>Loading...</div>}
    </ul>
  );
};

export default BooksContainer;
