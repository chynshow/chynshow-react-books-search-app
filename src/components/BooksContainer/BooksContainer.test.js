import React from 'react';
import BooksContainer from './';
import { mount } from 'enzyme';

it('should renders BooksContainer component', () => {
  const c = mount(<BooksContainer />);
  expect(c).toMatchSnapshot();
});
