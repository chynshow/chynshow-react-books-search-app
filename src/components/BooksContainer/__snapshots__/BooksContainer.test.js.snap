// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should renders BooksContainer component 1`] = `
<BooksContainer>
  <ul
    className="l-book-container"
  >
    <ScrollUp
      duration={250}
      easing="easeOutCubic"
      showUnder={160}
      style={
        Object {
          "bottom": 50,
          "cursor": "pointer",
          "position": "fixed",
          "right": 30,
          "transitionDelay": "0s",
          "transitionDuration": "0.2s",
          "transitionTimingFunction": "linear",
        }
      }
      topPosition={0}
    >
      <div
        onClick={[Function]}
        style={
          Object {
            "bottom": 50,
            "cursor": "pointer",
            "opacity": 0,
            "position": "fixed",
            "right": 30,
            "transitionDelay": "0s",
            "transitionDuration": "0.2s",
            "transitionProperty": "opacity, visibility",
            "transitionTimingFunction": "linear",
            "visibility": "hidden",
          }
        }
      >
        <button
          className="c-btn c-btn--primary c-scroll-btn"
        >
          <ChevronUpSVG>
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-chevron-up fa-w-14"
              data-icon="chevron-up"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M240.971 130.524l194.343 194.343c9.373 9.373 9.373 24.569 0 33.941l-22.667 22.667c-9.357 9.357-24.522 9.375-33.901.04L224 227.495 69.255 381.516c-9.379 9.335-24.544 9.317-33.901-.04l-22.667-22.667c-9.373-9.373-9.373-24.569 0-33.941L207.03 130.525c9.372-9.373 24.568-9.373 33.941-.001z"
                fill="currentColor"
              />
            </svg>
          </ChevronUpSVG>
        </button>
      </div>
    </ScrollUp>
  </ul>
</BooksContainer>
`;
